src/mlpack/bindings/go/print_go.cpp

```diff
@@ -0,0 +1,298 @@
+    }
+  }
+
+  for (ParamIter it = parameters.begin(); it != parameters.end(); ++it)
+  {
+    const util::ParamData& d = it->second;
+    if (d.input && !d.required &&
+        d.name != "help" && d.name != "info" &&
+        d.name != "version")
+      inputOptions.push_back(it->first);
+  }
+
+  // First, we must generate the mlpack package name.
+  cout << "package mlpack" << endl;
+  cout << endl;
+
+  // Now we must print the cgo's import libraries and files.
+  cout << "/*" << endl;
+  cout << "#cgo CFLAGS: -I/capi -I. -g -Wall" << endl;
+  cout << "#cgo LDFLAGS: -L/usr/local/lib/ -lm "
+      << "-lmlpack -L. -lmlpack_go_" << functionName << endl;
+  cout << "#include <capi/" << functionName << ".h>" << endl;
+  cout << "#include <stdlib.h>" << endl;
+  cout << "*/" << endl;
+  cout << "import \"C\" " << endl;
+  cout << endl;
+
+  // Then we must print the import of the gonum package.
+  cout << "import (" << endl;
+  cout << "  " << "\"gonum.org/v1/gonum/mat\" " << endl;
+  for (size_t i = 0; i < inputOptions.size(); ++i)
+  {
+    const util::ParamData& d = parameters.at(inputOptions[i]);
+    size_t indent = 2;
+    CLI::GetSingleton().functionMap[d.tname]["ImportDecl"](d,
+        (void*) &indent, NULL);
+  }
+  cout << ")" << endl;
+  cout << endl;
+
+  // Capitalize the first letter of parameter name so it is
+  // of exported type in Go.
+  std::string goFunctionName = functionName;
+  goFunctionName[0] = std::toupper(goFunctionName[0]);
+
+  // Print Go method configuration struct
+  cout << "type " << goFunctionName << "OptionalParam struct {"
+      << std::endl;
+  for (size_t i = 0; i < inputOptions.size(); ++i)
+  {
+    const util::ParamData& d = parameters.at(inputOptions[i]);
+    size_t indent = 4;
+    CLI::GetSingleton().functionMap[d.tname]["PrintMethodConfig"](d,
+        (void*) &indent, NULL);
+  }
+  cout << "}" << endl;
+  cout << endl;
+
+  // Print Go method configurate struct initialization
+  cout << "func Initialize" << goFunctionName << "() *"
+      << goFunctionName << "OptionalParam {"
+      << endl;
+  cout << "  " << "return &" << goFunctionName << "OptionalParam{" << endl;
+  for (size_t i = 0; i < inputOptions.size(); ++i)
+  {
+    const util::ParamData& d = parameters.at(inputOptions[i]);
+    size_t indent = 4;
+    CLI::GetSingleton().functionMap[d.tname]["PrintMethodInit"](d,
+        (void*) &indent, NULL);
+
+  }
+  cout << "  " << "}" << endl;
+  cout << "}" << endl;
+  cout << endl;
+
+  // Then we must print utility function for model type parameters if needed.
+  for (ParamIter it = parameters.begin(); it != parameters.end(); ++it)
+  {
+    const util::ParamData& d = it->second;
+    if (d.input)
+      CLI::GetSingleton().functionMap[d.tname]["PrintModelUtilGo"](d, NULL, NULL);
+  }
+
+  // Print the comment describing the function and its parameters.
+  cout << "/*" << endl;
+  cout << "  " << HyphenateString(programInfo.documentation(), 2) << endl;
```

Comment:
It's really awesome that the documentation printing works great!  That is not an easy thing given how different it is than the Python printing.  I did find one slight bug though; in the perceptron documentation, we get this:

```
  >>> param := InitializePerceptron()
  >>> param.Training = training_data
  >>> param.Labels = training_labels
  >>> perceptron := Perceptron(param)
~~
  Then, this model can be re-used for classification on the test data test_data.
   The example below does precisely that, saving the predicted classes to
  predictions.
~~
  >>> param := InitializePerceptron()
  >>> param.Input_model = perceptron
  >>> param.Test = test_data
  >>> predictions := Perceptron(param)
```

But I think, we also have to ignore the non-used return values, so it should be `_, perceptron := Perceptron(param)` then `predictions, _ := Perceptron(param)`.  I think it should not be too hard, just need to loop over the output options and see which ones aren't being used when printing the output documentation.  (Maybe it is harder than I think.  I am not sure, I haven't thought very hard about it...)

In any case the autogenerated documentation looks great otherwise!  That is one of the trickiest parts I had a lot of problems with for Python.  :rocket: * :+1: * :tada: